llvm/lib/IR/Attributes.cpp

```diff
@@ -191,6 +192,36 @@ Attribute Attribute::get(LLVMContext &Context, Attribute::AttrKind Kind,
   return Attribute(PA);
 }
 
+Attribute Attribute::get(LLVMContext &Context, Attribute::AttrKind Kind,
+                         ArrayRef<ConstantRange> Val) {
+  assert(Attribute::isConstantRangeListAttrKind(Kind) &&
+         "Not a ConstantRangeList attribute");
+  LLVMContextImpl *pImpl = Context.pImpl;
+  FoldingSetNodeID ID;
+  ID.AddInteger(Kind);
+  ID.AddInteger(Val.size());
+  for (auto &CR : Val) {
+    CR.getLower().Profile(ID);
+    CR.getUpper().Profile(ID);
+  }
+
+  void *InsertPoint;
+  AttributeImpl *PA = pImpl->AttrsSet.FindNodeOrInsertPos(ID, InsertPoint);
+
+  if (!PA) {
+    // If we didn't find any existing attributes of the same shape then create a
+    // new one and insert it.
+    void *Mem = pImpl->Alloc.Allocate(
+        ConstantRangeListAttributeImpl::totalSizeToAlloc(Val),
+        alignof(ConstantRangeListAttributeImpl));
+    PA = new (Mem) ConstantRangeListAttributeImpl(Kind, Val);
+    pImpl->AttrsSet.InsertNode(PA, InsertPoint);
```

Comment:
I guess this ends up working fine right now where the ConstantRanges have <= 64 bits, but in general I think this may result in a memory leak due to the dtor not running.

Reply:
How to call the dtor explicitly?

Reply:
hmm I'm not seeing where the existing ConstantRangeAttributeImpl's destructor gets called

Reply:
ah this should be using a separate `SpecificBumpPtrAllocator` like `ConstantRangeAttributeAlloc` so that we make sure to call destructors when the LLVMContext is destroyed

Reply:
oh thanks for this info! Changed to `SpecificBumpPtrAllocator`. Please take another look :-)